Update holidays to 0.26

arkid15r opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Bump holidays version from 0.21.13 to the latest 0.26.  Add holidays localization support.

The release notes:
  - https://github.com/dr-prodigy/python-holidays/releases/tag/v.0.26
  - https://github.com/dr-prodigy/python-holidays/releases/tag/v.0.25
  - https://github.com/dr-prodigy/python-holidays/releases/tag/v.0.24
  - https://github.com/dr-prodigy/python-holidays/releases/tag/v.0.23
  - https://github.com/dr-prodigy/python-holidays/releases/tag/v.0.22


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

arkid15r committed:
Bump holidays package version from 0.21.13 to 0.24
Fix Germany holidays tests (l10n).

arkid15r committed:
Bump holidays package version from 0.21.13 to 0.24
Fix Germany holidays tests (l10n).

arkid15r committed:
Merge branch 'dev' into bump-holidays-version-0.24

gjohansson-ST commented on tests/components/workday/test_config_flow.py:
```diff
@@ -346,7 +346,7 @@ async def test_form_incorrect_dates(hass: HomeAssistant) -> None:
             CONF_WORKDAYS: DEFAULT_WORKDAYS,
             CONF_ADD_HOLIDAYS: ["2022-12-12"],
-            CONF_REMOVE_HOLIDAYS: ["Weihnachtstag"],
+            CONF_REMOVE_HOLIDAYS: ["Christmas"],
```
This would be quite a breaking change if you can not express removal of holidays based on local language anymore?

gjohansson-ST replied:
Not sure I get it but the configuration should still be in local language regardless of system locale.
For example my own system is in English but I'm still Swedish and would like to write "Julafton" and not "Christmas"
Please change it back

arkid15r replied:
>For example my own system is in English but I'm still Swedish and would like to write "Julafton" and not "Christmas"

Thank you for the use case example.

>Please change it back

I don't think it works like that. The current python-holidays l10n changes the default holiday names output for some countries. So the component code needs to be adapted. That's what I meant by 

>// I guess more changes may be required to deal with the localized holidays.

I'll look into what can be done in order to integrate the new changes smoothly based on the example you provided.

gjohansson-ST replied:
Ideally would be to be able to use both English and localized names.

gjohansson-ST replied:
Any further progress here?

arkid15r reviewed (commented):
Hi @gjohansson-ST, thanks for the quick review!

arkid15r committed:
Merge branch 'dev' of github.com:arkid15r/home-assistant-core into bump-holidays-version-0.24

arkid15r committed:
Implement language option support

  - Bump Python Holidays version to 0.26
  - Add language validation (rework province validation a bit)
  - Add language to the config flows (move province seelct field
    to the schema's top and add the language select field right
    below it)
  - Remove stale dependencies
  - Update strings.json
  - Add/update tests

arkid15r committed:
Implement language option support

  - Bump Python Holidays version to 0.26
  - Add language validation (rework province validation a bit)
  - Add language to the config flows (move province seelct field
    to the schema's top and add the language select field right
    below it)
  - Remove stale dependencies
  - Update strings.json
  - Add/update tests

arkid15r committed:
Add a missing key

arkid15r committed:
Add a missing key

arkid15r committed:
Merge branch 'bump-holidays-version-0.24' of github.com:arkid15r/home-assistant-core into bump-holidays-version-0.24

arkid15r committed:
Update code style

arkid15r commented:
Hey @gjohansson-ST, 
here is an implementation of language support option for workday sensor holidays. 

As it's been my first time experience with config flow please review thoroughly. I bet there is some excessive stuff and/or missing one too.

Thank you!

gjohansson-ST commented:
I don't think we should ask for the language. We should just use the local language eg German for Germany, Swedish for Sweden and so on.

arkid15r commented:
> I don't think we should ask for the language. We should just use the local language eg German for Germany, Swedish for Sweden and so on.

I'm sure that wouldn't be hard to implement. 

First of all, I'd like to note that the default/local country's language (if any) has already been used as a default for the new Language field.

I'd also like to point out that as a Home Assistant user I'd benefit more from an ability to select one of the available localizations for a country holidays instead of using non-configurable default value set behind the scenes. 

 A couple of use case examples:
  - some countries have multiple official languages, e.g., Canada where I expect people would want to be able to pick their preferred choice
  - when I create a Workday sensor for a country w/ the default language I don't know I'd definitely prefer using one of localizations I'm familiar with instead of the default one (e.g., use en_US or uk instead of sv for Sweden holidays addition/exclusion)

I don't mind simplifying the implementation based on your comment if that's the best/optimal UX for HA users. Could you elaborate on your recommendation a bit -- I'd like to hear more on why in your opinion we shouldn't ask for the language.

Thank you!

arkid15r committed:
Merge branch 'bump-holidays-version-0.24' of github.com:arkid15r/home-assistant-core into bump-holidays-version-0.24

arkid15r committed:
Merge branch 'dev' into bump-holidays-version-0.24

gjohansson-ST commented:
@arkid15r Maybe it's better to open a new PR given there is now many people subscribed to this PR from your merge (you should rebase instead).
Secondly it should still mimic current behavior. Implementation of language option is not bad but should be an option then and in a follow-up PR

frenck commented:
I've marked this PR, as the CI is failing which needs to be fixed. Please add the missing test coverage. 

Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

arkid15r commented:
> @arkid15r Maybe it's better to open a new PR given there is now many people subscribed to this PR from your merge (you should rebase instead). Secondly it should still mimic current behavior. Implementation of language option is not bad but should be an option then and in a follow-up PR

@gjohansson-ST I'm closing this out in favor of https://github.com/home-assistant/core/pull/95091
Please review the PR with actionable comments if any.

> ... the CI is failing which needs to be fixed. Please add the missing test coverage.
> 

@frenck Thanks for pointing out an issue. I understand the proper test coverage concern.
